Fixes #34093 - Don't eager load all authorized resources

tbrisker opened the pull request:
The authorizer cache currently loads all permitted resources to memory
when `.includes?` is called just to check if one resource is in the
list. This can get very heavy when authorizing resources that have many
records, for example ConfigReports.
Instead of loading the whole scope to memory, we use a fast SQL query
to check for the specific subject's inclusion in the authorized scope.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

tbrisker committed:
Fixes #34093 - Don't eager load all authorized resources

The authorizer cache currently loads all permitted resources to memory
when `.includes?` is called just to check if one resource is in the
list. This can get very heavy when authorizing resources that have many
records, for example ConfigReports.
Instead of caching the scope and loading all AR objects to memory, we
can cache just the ids to reduce the memory usage of the cache. This is
still sub-optimal for resources that have millions of records, but is
faster for pages such as index pages which perform multiple permissions
checks for every item and don't need to perform an additional sql query
for every check.

ekohl reviewed (commented):
Does this change it from an in-memory lookup to a dynamic query or is it still cached?

tbrisker commented:
@ekohl that was my first though as well, but you'll notice that even before this change, what we are caching is the scope, for example:
```
 @cache[subject.class.to_s][permission].class
=> Host::Managed::ActiveRecord_Relation
```
so in fact every repeat call to this method, even with the same parameters, was triggering another full load of the scope just to check if the subject is included in the array. `.include?` triggers the scope load into an array, but the cache doesn't cache this resulting array. 
On the other hand, now with this change, rails can actually automatically cache repeat calls that generate an identical sql query so we don't even hit the DB on repeated queries with the same parameters.

tbrisker commented on app/services/authorizer_cache.rb:
```diff
@@ -9,6 +9,6 @@ def collection_cache_lookup(subject, permission)
     collection = @cache[subject.class.to_s][permission] ||=
       find_collection(subject.class, :permission => permission)
```
`find_collection` is a very expensive function call, and it's worth caching it when we can.

ekohl reviewed (dismissed):
Should this use `exists?` too so it's consistent?
https://github.com/theforeman/foreman/blob/96a6076fb39fe5800481c3a6e2467cf66d062437/app/services/authorizer.rb#L22-L23

It's a bit confusing to me that `collection_cache_lookup` also does the validation instead of being a pure cache for `find_collection`. Looking at it fresh, I'd have expected something like this in `Authorizer`:
```ruby
collection = cache ? collection_cache_lookup(subject, permission) : find_collection(subject.class, :permission => permission)
collection.exists?(subject.id)
```

However, your current patch can be safely cherry picked.

ezr-ondrej commented:
I'd agree that the caching should be the same just with saving or not, @tbrisker could you open a PR unifying it as @ekohl points out?

tbrisker commented:
Actually, looking at it again looks like i am wrong, the collection gets marked as `loaded` after the first `.include?` call, so additional calls don't trigger additional queries.  However, I still think it is better to make multiple queries for the same collection rather than materialize the whole scope in memory when in almost all cases we don't actually care about more than $per_page records from the collection.
If we want to keep the scope results in memory, a middle ground might be to `pluck(:id)` so at least we have an array of ids in memory instead of potentially thousands or more of ActiveRecord objects

ekohl commented:
The thing I do wonder about is about instances changing. If the cache survives between requests, will `include?` actually match if rows have been changed? If so, that could lead to subtle bugs. With IDs that wouldn't be an issue since generally speaking they don't change.

tbrisker commented:
@ekohl afaik the instance variables get reset for every request, this is only for caching multiple checks in the same request

tbrisker commented:
to sum up the three options:

| Option | Pros | cons |
|-|-|-|
| 1. Current implementation | scope loaded once to memory, it is known to work | High memory load for large scopes |
| 2. Cache scope + exists? | Lowest memory use | Every check triggers another sql query (identical checks will get cached by rails) |
| 3. Cache IDs | Lower memory usage then current. Likely faster for pages that performs many checks | Could still get high memory load for large scopes. Array scanning not optimal for large arrays

So I'm thinking in most cases for the UI option 2 would be the best, triggering a few fast sql queries will likely not impact performance as much as loading a lot of records to memory. 
But for smaller scopes or for api calls that may trigger many permissions checks, option 3 would likely be fastest.

We could try some more complex caching mechanism that tries to figure out if we should use option 2 or 3, but that would add complexity that i'm not sure would be worth the potential bugs it would introduce or the performance gain.

I'd be happy to hear your opinions on which one we should implement.

ekohl commented:
For option 2 https://guides.rubyonrails.org/caching_with_rails.html#sql-caching suggests it doesn't really matter in which order queries come. Both the AuthorizerCache and Rails SQL cache are scoped to a single request. It certainly requires a comment that it in the code that we rely on SQL cache. However, it does mean that if you check it for multiple IDs you'll issue multiple queries.

Option 3 will perform a lot better on loops over collections. So if you display a list of records and check for every entry if the user is authorized. I don't know how many of those we have.

Essentially the question is: do we loop over collections? And how much slower is it to send multiple queries vs loading a (potentially) huge list. Worst case you need to (say) load a million IDs and check against that. If you only check it once that will obviously be a downside. With twice it may also be faster to load 2 queries. However, with 50 or 100 the round trip may be more overhead. Especially with a remote DB where network latency is involved.

Any idea if we find out the common usage patterns?

lzap commented:
> @ekohl afaik the instance variables get reset for every request

Can confirm, if we shared this across request without thread locking, it would be a disaster :-)

What is the reason to hold the whole collection in the memory? When it comes to caching, I actually want to only cache the result, that is a boolean. Takes just 4 bytes instead of an instance or array of something. As noted, Rails will cache some of the results called in `find_collection` so it is okay to call this multiple times but honestly, that method is a candidate for low-level caching - it is pretty beefy and permissions do not change that often (and when they do it is probably okay to wait 5 minutes until things get cleared out).

Please consider adding https://community.theforeman.org/t/foreman-telemetry-api-for-developers/26409 in a separate commit for this, I think permission system needs more performance focus and we need numbers.

tbrisker commented:
@ekohl I believe in normal UI usage we shouldn't get more than 20 or maybe 50 checks per page, which i think would be fine doing sql queries (these are all id lookups, so they shouldn't take more then 1ms to resolve - although external db might add latency). 
I was expecting the API to be the main problem, since there we sometimes load 1000 records in a single request, but then found this comment:
https://github.com/theforeman/foreman/blob/7114ed0d86d4a3b7b366b0a0941564e62fc0765e/app/views/api/v2/layouts/permissions.json.rabl#L6-L7

which suggests that we don't actually call `.can?` in the api anyways exactly for this reason. There is one use of `.can?` in the host show api, but that is only for single hosts. API controllers use different authorization methods to scope the records for display and do not seem to reach this code path.

ekohl commented:
> What is the reason to hold the whole collection in the memory? When it comes to caching, I actually want to only cache the result, that is a boolean. Takes just 4 bytes instead of an instance or array of something.

Let's say you have (in pseudo-code):
```ruby
a_list_of_instances.each do |instance|
  if authorized_for(instance)
    ...
  end
end
```

Then having the full list is faster.

lzap commented:
I think this kind of caching makes zero sense to me given the price we need to pay in terms of memory. Tomer is right, we only call this dozens of times. Here are some examples:

Host index page with three hosts (in practice we default to 20 hosts per page):

```
2021-12-06T16:24:24 [W|app|6b75bbd8] CAN? create_bookmarks
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? edit_hosts Host::Managed 28
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? destroy_hosts Host::Managed 28
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? edit_hosts Host::Managed 29
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? destroy_hosts Host::Managed 29
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? edit_hosts Host::Managed 2
2021-12-06T16:24:30 [W|app|0e76b1d9] CAN? destroy_hosts Host::Managed 2
2021-12-06T16:24:31 [W|app|0e76b1d9] CAN? create_bookmarks
```

Opening a host detail page:

```
2021-12-06T16:25:53 [W|app|c92e9fa0] CAN? cockpit_hosts Host::Managed 2
2021-12-06T16:25:53 [W|app|c92e9fa0] CAN? edit_hosts Host::Managed 2
2021-12-06T16:25:53 [W|app|c92e9fa0] CAN? create_hosts Host::Managed 2
2021-12-06T16:25:53 [W|app|c92e9fa0] CAN? build_hosts Host::Managed 2
2021-12-06T16:25:53 [W|app|c92e9fa0] CAN? destroy_hosts Host::Managed 2
2021-12-06T16:25:54 [W|app|53a6168e] CAN? edit_subnets Subnet::Ipv4 1
```

Listing all reports (host reports):

```
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 169
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 169
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 167
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 167
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 164
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 164
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 166
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 166
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 165
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 165
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 163
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 163
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 162
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 162
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 160
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 160
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 161
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 161
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 158
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 158
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 159
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 159
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 157
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 157
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 156
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 156
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? edit_host_reports HostReport 168
2021-12-06T16:26:49 [W|app|dd57eba6] CAN? destroy_host_reports HostReport 168
```

Opening an audit index page:

```
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 41
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 41
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 41
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 41
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 40
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_locations Location 2
2021-12-06T16:28:16 [W|app|69776b0c] CAN? edit_organizations Organization 1
2021-12-06T16:28:16 [W|app|69776b0c] CAN? view_discovered_hosts Host::Discovered 40
```

Provisioning templates:

```
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 43
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 30
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 10
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 60
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 61
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 62
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 63
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 52
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 53
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 37
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 11
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 105
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 54
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 64
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 65
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 137
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 136
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 138
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 135
2021-12-06T16:30:25 [W|app|03907dd6] CAN? edit_provisioning_templates ProvisioningTemplate 66
2021-12-06T16:30:25 [W|app|03907dd6] CAN? create_bookmarks
```

Obviously showing detail of any resource, editing a resource or deleting it generated very small amount of calls (less then 10) for each request.

Conclusion: This caching does not bring much value, at least in the current implementation. I suggest to completely remove it. There might be some (batch) processing triggering many these calls tho, there was probably a reason why this was addded. @ares

lzap commented:
Btw git blame shows: 1ff7e25908f81912458a0fd85c421fb5feeb90a1

tbrisker commented:
@lzap I believe the cache is still worth it so we don't call `find_collection` every time we call `can`, since that method is very expensive (it builds a scope that checks for user permissions and permitted taxonomies). The only question is if we should run 20 queries to check that the resource is permitted or load all permitted resource(_id)s to memory and check the cached list once

ekohl commented:
Perhaps we need 2 APIs: one for bulk access (like our REST API and Provisioning Templates) and one for single instance.

Having said that: https://community.theforeman.org/t/high-memory-usage-of-foreman-2-5-4/26410/18 suggests that what I said in https://github.com/theforeman/foreman/pull/8972#issuecomment-986843683 is pretty much true and in favor of option 2 for our common use cases.

So I think we should merge the current patch and then see if we can implement a more generic bulk filter API.

tbrisker committed:
Refs #34093 - Minor authorizer clean ups

- Use `.exists?` instead of `.where.any?` and `.where.present?`.
`.any?` avoids triggering another sql query for loaded relations, but in
this specific case the relation won't be loaded. `.present?` loads the
relation to an array and checks if it is empty, which is obviously worse
then checking in sql.
- Use regular `Hash` instead of `HashWithIndefferentAccess`
HWIA comes with a performance overhead since it needs to handle both
string and symbol keys. Since the `@cache` is only used in one place, we
can control the keys and use a regular Hash instead.

ezr-ondrej commented:
I'm in favor of option 2 and merging this, we need to fix tests tho

lzap commented:
I get it now, right. Yeah, `pluck` approach might have sense for some bulk API that could pre-authorize, however, we would need to find a proper use-case for this method before starting looking into this. I haven't see one yet. The patch looks good, most importantly it solved the issue for the reporter.

ezr-ondrej commented:
Rubocop doesn't like `Hash.new` over `{}` :)

tbrisker commented:
Fixed the tests in the first commit and added a commit with a few small cleanups around this method call.

lzap commented on app/services/authorizer_cache.rb:
```diff
@@ -1,14 +1,13 @@
 
   def collection_cache_lookup(subject, permission)
-    collection = @cache[subject.class.to_s][permission] ||=
+    collection = @cache[subject.class.to_s][permission.to_s] ||=
```
I was thinking the same, indifferent access is often used for no reason. It could be even flattened to avoid creating many Hash objects, however, probably not worth it.

lzap reviewed (dismissed):
Nice one.

tbrisker commented:
When applying this patch, it [appears](https://community.theforeman.org/t/high-memory-usage-of-foreman-2-5-4/26410/23) that the loading the host index page with 100 hosts is slower by about 70% and causes about 85% more memory allocations. I'm not sure why that occurs, but it might be due to the additional overhead of running a few hundred additional sql queries (there are about 3 permission checks per row in the host index page).

ekohl reviewed (commented):
Now I'm wondering if we should actually do it right. Effectively we can have 3 levels of caching:
* Nothing
* Cache the collection but not the results (option 2)
* Cache the results (option 3)

So should we introduce a `cache_results` parameter to `collection_cache_lookup`?

Then the case where we can pull in a lot of resources, we want to not cache the results. However, this is probably hard to automatically determine (you essentially need a count which negates all the positive effects).

jturel commented:
Love seeing the trend in resolving issues like this. It's nice when users are so willing to cooperate in the investigation too!

ekohl commented:
@tbrisker given that option 3 is a sure improvement over the current implementation without an increase in queries, should we start with that and then open an issue to track further improvements?

tbrisker committed:
Refs #34093 - Allow skipping cache from `User.can?`

tbrisker committed:
Refs #34093 - Improve host config status performance

- Skip the authorizer cache. There can be millions of ConfigReports in
  the db, there is no need to load all of them just to check if the user
  can see the last one.
- Cache the status_link method result. The status_link method checks for
  the user's permissions and is called 3 times - twice in
  host_detailed_status_list directly and once inside
  host_global_status_icon_class. We can avoid the extra permissions
  check by caching the method's results.

tbrisker commented:
Updated the cache to save IDs, and modified the specific check in the overview pane to skip the cache. This should improve performance overall for pages that actually need the cache (such as the host index page), while not loading all config reports to memory just to check if the user can see the latest one.

ekohl reviewed (dismissed):
Thanks!

lzap commented on app/models/host_status/configuration_status.rb:
```diff
@@ -109,10 +109,11 @@ def self.bit_mask(config_status)
 
-      last_report && Rails.application.routes.url_helpers.config_report_path(last_report)
+      @config_status_link = last_report && Rails.application.routes.url_helpers.config_report_path(last_report)
     end
```
View code in a model. Foreman in 2021 :-)

I like the solution.

tbrisker replied:
this code should definitely move to a helper :)

lzap commented on app/models/user.rb:
```diff
@@ -175,12 +175,12 @@ def check_permissions_after_save
       @authorizer ||= Authorizer.new(self)
-      @authorizer.can?(permission, subject)
+      @authorizer.can?(permission, subject, cache)
     end
```
Maybe you could actually drop a comment here and in the authorizer about the cache argument - why it is there and how to set it (when calling on data that can possibly load too many items in the cache).

lzap commented on app/services/authorizer_cache.rb:
```diff
@@ -1,14 +1,14 @@
+      find_collection(subject.class, :permission => permission).pluck(:id)
 
-    collection.include?(subject)
+    collection.include?(subject.id)
```
Can you add a warning when too many items are loaded into memory? It can have a very meaningful explanation and big impact on how we are able to find these bottlenecks. Something like:

Authorizer cache loaded more than 1000 items, consider calling can? without cache to improve memory consumption.

I think this should be part of the patch which will backport, it will allow us to see more of these hotspots.

tbrisker replied:
good idea, i'll add a log but i think the threshold should be 100k - otherwise any user with many hosts will see this log all the time, and loading a 100k integer array is just 400k of memory used.

lzap replied:
Nice yeah 1000 was definitely too low, this looks good.

lzap replied:
Oh wait, this will issue INFO log on every check, I thought to have it in the block where `find_collection` is actually called to show it just once per subject check. Also consider a DEBUG level, although I think INFO gives better visibility. Since 100k is pretty high, we could actually do even WARN.

tbrisker replied:
We don't know the collection size inside `find_collection`, we only load the records after we get the collection.

lzap replied:
Sure but you are a programmer, you can solve this! :-)

ekohl replied:
Does `.size` create a separate query that uses `COUNT()` or does it use the resulting set? Or does `.pluck()` return a simple array?

lzap replied:
Pluck return an array.

lzap reviewed (dismissed):
I have some comments, tho, I do not want to block this.

tbrisker committed:
Refs #34093 - Document Authorizer#can? and log huge cache loads

Also flatten the authorizer cache instead of a nested hash.

tbrisker commented:
Added comments and logging line, tests were previously green.

tbrisker commented:
Updated the logging to only run once, and flattened the cache hash for cleaner code.

lzap reviewed (approved):
I have no further comments, looks great and this can possibly find other places we have in the codebase.

ezr-ondrej reviewed (approved):
Thanks @tbrisker ! :+1:

ezr-ondrej commented:
CP 3.1: b0fe40335 (last commit, see parents for the previous ones)
